File: source/src/main/java/com/aerospike/examples/InteractiveMenu.java
package com.aerospike.examples;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import com.aerospike.client.IAerospikeClient;
import com.aerospike.mapper.tools.AeroMapper;

/**
 * Handles the interactive menu system for use case selection and execution.
 * This class follows the Single Responsibility Principle by only handling menu interaction.
 */
public class InteractiveMenu {
    

    
    private final IAerospikeClient client;
    private final AeroMapper mapper;
    private final UseCaseExecutor executor;
    
    /** Currently filtered use cases for search results */
    private List<UseCase> filteredUseCases;
    
    /** Current search term for highlighting */
    private String currentSearchTerm;
    
    /** Whether current search is using regex */
    private boolean isRegexSearch;
    
    public InteractiveMenu(IAerospikeClient client, AeroMapper mapper) {
        this.client = client;
        this.mapper = mapper;
        this.executor = new UseCaseExecutor(client, mapper);
        this.filteredUseCases = new ArrayList<>(UseCaseRegistry.getAllUseCases());
    }
    
    /**
     * Run the interactive menu system.
     * @param width The terminal width for formatting
     */
    public void runMenu(int width) {
        boolean summary = false;
        boolean skipUseCaseDetails = false;
        
        while (true) {
            showStartHeader();
            if (!skipUseCaseDetails) {
                showUseCaseDetails(width, summary);
            }
            skipUseCaseDetails = false;
            
            // Show search status if active
            if (currentSearchTerm != null && !currentSearchTerm.trim().isEmpty()) {
                System.out.println(AnsiColors.YELLOW + "Search active: " + 
                    (isRegexSearch ? "regex '" : "'") + currentSearchTerm + "' " +
                    "(" + filteredUseCases.size() + " results)" + AnsiColors.RESET);
            }
            
            System.out.printf("Enter a number to run a use case, or enter a command (help for commands): ");
            Scanner scanner = new Scanner(System.in);
            String input = scanner.nextLine().trim();
            
            if (input.isEmpty()) {
                continue;
            }
            
            // Handle commands
            if (input.equalsIgnoreCase("help") || input.equalsIgnoreCase("h") || input.equals("?")) {
                showHelp();
                skipUseCaseDetails = true;
                continue;
            }
            
            if (input.equalsIgnoreCase("summary")) {
                summary = true;
                continue;
            }
            
            if (input.equalsIgnoreCase("full") || input.equalsIgnoreCase("details")) {
                summary = false;
                continue;
            }
            
            // Handle search commands
            if (input.startsWith("search ")) {
                String searchTerm = input.substring(7).trim();
                searchUseCases(searchTerm, false);
                continue;
            }
            
            if (input.startsWith("s ")) {
                String searchTerm = input.substring(2).trim();
                searchUseCases(searchTerm, false);
                continue;
            }
            
            if (input.startsWith("/")) {
                String searchTerm = input.substring(1).trim();
                searchUseCases(searchTerm, true);
                continue;
            }
            
            if (input.equalsIgnoreCase("clear") || input.equalsIgnoreCase("c")) {
                searchUseCases(null, false);
                continue;
            }
            
            if (input.equalsIgnoreCase("exit") || input.equalsIgnoreCase("quit")) {
                System.out.println("Goodbye!");
                break;
            }
            
            // Handle use case selection
            try {
                int selection = Integer.parseInt(input);
                if (selection > 0 && selection <= filteredUseCases.size()) {
                    invokeUseCase(selection);
                } else {
                    System.out.println(AnsiColors.RED + "Invalid selection. Please enter a number between 1 and " + 
                        filteredUseCases.size() + "." + AnsiColors.RESET);
                }
            } catch (NumberFormatException e) {
                System.out.println(AnsiColors.RED + "Invalid input. Please enter a number or a command (type 'help' for available commands)." + AnsiColors.RESET);
            }
        }
    }
    
    /**
     * Show the start header.
     */
    private void showStartHeader() {
        System.out.println();
        System.out.println(AnsiColors.BOLD + "Aerospike Use Case Cookbook" + AnsiColors.RESET);
        System.out.println("==============================");
        System.out.println();
    }

    private String formatColors(String str, String color) {
        return color + highlightSearchTerms(str, color) + AnsiColors.RESET;
    }
    private String formatLine(String indexStr, String nameStr, String descStr, String format) {
        return String.format("| %s | %s | %s |", 
                formatColors(indexStr, format), 
                formatColors(nameStr, format), 
                formatColors(descStr, format));
    }
    
    private String formTagsString(String[] tags, String color) {
        StringBuffer sb = new StringBuffer().append("Tags: ");
        for (String tag : tags) {
            if (sb.length() > 0) {
                sb.append(' ');
            }
            sb.append(AnsiColors.REVERSE).append(tag).append(AnsiColors.RESET).append(color);
        }
        return sb.toString();
    }
    
    /**
     * Show use case details in a formatted table.
     * @param width The terminal width
     * @param summary Whether to show summary or full details
     */
    private void showUseCaseDetails(int width, boolean summary) {
        if (filteredUseCases.isEmpty()) {
            System.out.println(AnsiColors.RED + "No use cases found matching your search criteria." + AnsiColors.RESET);
            return;
        }
        
        final int columnSpaces = 6;
        final int columnChars = 4;
        int longestName = getLongestUseCaseName();
        int indexWidth = String.valueOf(filteredUseCases.size()).length();
        int nameWidth = Math.min(longestName, 50);
        int descWidth = width - indexWidth - nameWidth - columnChars - columnSpaces;
        
        String indexHeader = padCenter("#", indexWidth);
        String nameHeader = padCenter("Name", nameWidth);
        String descHeader = padCenter(summary ? "Summary" : "Description", descWidth);
        
        String horizontalLine = repeat("-", width);
        System.out.println(horizontalLine);
        System.out.println(formatLine(indexHeader, nameHeader, descHeader, AnsiColors.BOLD));
        
        System.out.println(horizontalLine);
        
        for (int i = 0; i < filteredUseCases.size(); i++) {
            UseCase uc = filteredUseCases.get(i);
            String index = padLeft(String.valueOf(i + 1), indexWidth);
            String color = (i % 2 == 0) ? AnsiColors.COLOR2 : AnsiColors.COLOR1; // alternating colors
            String name = padRight(uc.getName(), nameWidth);
            List<String> wrappedDesc = wrapText(formUseCaseText(uc, summary), descWidth);
            
            for (int j = 0; j < wrappedDesc.size(); j++) {
                String descLine = padRight(wrappedDesc.get(j), descWidth);
                if (j == 0) {
                    System.out.println(formatLine(index, name, descLine, color));
                } else {
                    System.out.println(formatLine(
                            padLeft("", indexWidth),
                            padRight("", nameWidth),
                            descLine, color));
                }
            }
            if (uc.getReference() != null) {
                System.out.println(formatLine( 
                        padLeft("", indexWidth),
                        padRight("", nameWidth),
                        padRight("", descWidth),
                        color));
                System.out.println(formatLine( 
                        padLeft("", indexWidth),
                        padRight("", nameWidth),
                        padRight("See: " + uc.getReference(), descWidth),
                        color));
            }
            if (uc.getTags() != null && uc.getTags().length > 0) {
                System.out.println(formatLine(
                        padLeft("", indexWidth),
                        padRight("", nameWidth),
                        padRight(formTagsString(uc.getTags(), color), descWidth),
                        color));
            }
            System.out.println(horizontalLine);
        }
    }
    
    /**
     * Get the longest use case name for formatting.
     * @return The length of the longest name
     */
    private int getLongestUseCaseName() {
        return filteredUseCases.stream()
                .mapToInt(uc -> uc.getName().length())
                .max()
                .orElse(0);
    }
    
    /**
     * Format use case text for display.
     * @param uc The use case
     * @param summary Whether to show summary or full description
     * @return The formatted text
     */
    private String formUseCaseText(UseCase uc, boolean summary) {
        if (summary) {
            return uc.getDescription().split("\\.")[0] + ".";
        } else {
            return uc.getDescription();
        }
    }
    
    /**
     * Wrap text to fit within a specified width.
     * Preserves explicit newlines in the text.
     * @param text The text to wrap
     * @param width The maximum width
     * @return List of wrapped lines
     */
    private List<String> wrapText(String text, int width) {
        List<String> lines = new ArrayList<>();
        
        // First split by explicit newlines to preserve them
        String[] paragraphs = text.split("\\n");
        
        for (String paragraph : paragraphs) {
            // Then wrap each paragraph independently
            String[] words = paragraph.split("\\s+");
            StringBuilder currentLine = new StringBuilder();
            
            for (String word : words) {
                if (currentLine.length() + word.length() + 1 <= width) {
                    if (currentLine.length() > 0) {
                        currentLine.append(" ");
                    }
                    currentLine.append(word);
                } else {
                    if (currentLine.length() > 0) {
                        lines.add(currentLine.toString());
                        currentLine = new StringBuilder(word);
                    } else {
                        // Word is longer than width, split it
                        lines.add(word.substring(0, width));
                        currentLine = new StringBuilder(word.substring(width));
                    }
                }
            }
            
            if (currentLine.length() > 0) {
                lines.add(currentLine.toString());
            }
        }
        
        return lines;
    }
    
    /**
     * Pad a string to the right with spaces.
     * @param text The text to pad
     * @param width The target width
     * @return The padded string
     */
    private String padRight(String text, int width) {
        int extraChars = 0;
        int startIndex = text.indexOf("\033");
        while (startIndex > 0) {
            int endIndex = text.indexOf('m', startIndex);
            extraChars += endIndex - startIndex + 1;
            startIndex = text.indexOf("\033", endIndex);
        }
        return String.format("%-" + (width+extraChars) + "s", text);
    }
    
    private String padCenter(String text, int width) {
        int strLength = text.length();
        int leftPadding = Math.max(0, (width - strLength)/2);
        int rightPadding = Math.max(0, width - leftPadding - strLength);
        return repeat(" ", leftPadding) + text + repeat(" ", rightPadding);
    }
    
    /**
     * Pad a string to the left with spaces.
     * @param text The text to pad
     * @param width The target width
     * @return The padded string
     */
    private String padLeft(String text, int width) {
        return String.format("%" + width + "s", text);
    }
    
    /**
     * Repeat a character a specified number of times.
     * @param ch The character to repeat
     * @param count The number of times to repeat
     * @return The repeated string
     */
    private String repeat(String ch, int count) {
        return ch.repeat(count);
    }
    
    /**
     * Highlight search terms in text.
     * @param text The text to highlight
     * @param preserveColor The color to preserve after highlighting
     * @return The highlighted text
     */
    private String highlightSearchTerms(String text, String preserveColor) {
        if (currentSearchTerm == null || currentSearchTerm.trim().isEmpty()) {
            return text;
        }
        
        try {
            Pattern pattern;
            if (isRegexSearch) {
                pattern = Pattern.compile(currentSearchTerm, Pattern.CASE_INSENSITIVE);
            } else {
                pattern = Pattern.compile(Pattern.quote(currentSearchTerm), Pattern.CASE_INSENSITIVE);
            }
            
            Matcher matcher = pattern.matcher(text);
            StringBuffer result = new StringBuffer();
            
            while (matcher.find()) {
                matcher.appendReplacement(result, AnsiColors.HIGHLIGHT + matcher.group() + AnsiColors.RESET + preserveColor);
            }
            matcher.appendTail(result);
            
            return result.toString();
        } catch (Exception e) {
            // If regex compilation fails, fall back to literal search
            return text.replaceAll("(?i)" + Pattern.quote(currentSearchTerm), 
                AnsiColors.HIGHLIGHT + "$0" + AnsiColors.RESET + preserveColor);
        }
    }
    
    /**
     * Search for use cases that match the given search term.
     * @param searchTerm The search term
     * @param useRegex Whether to use regex matching
     */
    private void searchUseCases(String searchTerm, boolean useRegex) {
        this.currentSearchTerm = searchTerm;
        this.isRegexSearch = useRegex;
        
        if (searchTerm == null || searchTerm.trim().isEmpty()) {
            // Clear search
            filteredUseCases = new ArrayList<>(UseCaseRegistry.getAllUseCases());
            return;
        }
        
        filteredUseCases = new ArrayList<>();
        Pattern pattern = null;
        
        try {
            if (useRegex) {
                pattern = Pattern.compile(searchTerm, Pattern.CASE_INSENSITIVE);
            } else {
                pattern = Pattern.compile(Pattern.quote(searchTerm), Pattern.CASE_INSENSITIVE);
            }
        } catch (Exception e) {
            // If regex compilation fails, fall back to literal search
            pattern = Pattern.compile(Pattern.quote(searchTerm), Pattern.CASE_INSENSITIVE);
        }
        
        for (UseCase uc : UseCaseRegistry.getAllUseCases()) {
            if (matchesSearch(uc, pattern)) {
                filteredUseCases.add(uc);
            }
        }
    }
    
    /**
     * Check if a use case matches the search pattern.
     * @param useCase The use case to check
     * @param pattern The search pattern
     * @return true if the use case matches
     */
    private boolean matchesSearch(UseCase useCase, Pattern pattern) {
        return pattern.matcher(useCase.getName()).find() ||
               pattern.matcher(useCase.getDescription()).find() ||
               pattern.matcher(useCase.getReference()).find() || 
               pattern.matcher(String.join(" ", useCase.getTags())).find();
    }
    
    /**
     * Invoke a use case by selection number.
     * @param selection The selection number (1-based)
     */
    private void invokeUseCase(int selection) {
        UseCase uc = filteredUseCases.get(selection - 1);
        executor.executeUseCase(uc, true, false, false); // true = interactive
    }
    
    /**
     * Display help information for commands.
     */
    private void showHelp() {
        System.out.println("Help commands:");
        System.out.println("  help           - Show this help text");
        System.out.println("Display commands:");
        System.out.println("  summary        - Show only summary (one line) use case descriptions");
        System.out.println("  full / details - Show detailed use case descriptions");
        System.out.println("Search commands:");
        System.out.println("  search <term>  - Search for term in names, descriptions, and URLs");
        System.out.println("  s <term>       - Short form of search");
        System.out.println("  /<regex>       - Search using regular expression");
        System.out.println("  clear          - Clear search and show all use cases");
        System.out.println("Running a use case:");
        System.out.println("  <shown number> - Execute the use case with the specified number");
        System.out.println("                  (If the use case has parameters, you'll be prompted to configure them)");
        System.out.println("Exiting the program:");
        System.out.println("  exit / quit    - Exit the program");
        System.out.println();
    }
} 

File: source/python-sdk/usecasecookbook/interactive_menu.py
"""Handles the interactive menu system for use case selection and execution.
Mirrors ../../java-sdk's InteractiveMenu - see ../java/README_SEARCH.md for the search
command reference, which applies unchanged here.
"""

import re
import shutil

from aerospike_sdk import SyncSession

from usecasecookbook import ansi_colors as c
from usecasecookbook import use_case_registry as registry
from usecasecookbook.use_case import UseCase
from usecasecookbook.use_case_executor import UseCaseExecutor


class InteractiveMenu:
    def __init__(self, session: SyncSession):
        self.executor = UseCaseExecutor(session)
        self.filtered_use_cases: list[UseCase] = list(registry.get_all_use_cases())
        self.current_search_term: str | None = None
        self.is_regex_search = False

    def run_menu(self, width: int | None = None) -> None:
        if width is None:
            width = shutil.get_terminal_size(fallback=(200, 24)).columns

        summary = False
        skip_use_case_details = False

        while True:
            self._show_start_header()
            if not skip_use_case_details:
                self._show_use_case_details(width, summary)
            skip_use_case_details = False

            if self.current_search_term and self.current_search_term.strip():
                kind = "regex '" if self.is_regex_search else "'"
                print(
                    f"{c.YELLOW}Search active: {kind}{self.current_search_term}' "
                    f"({len(self.filtered_use_cases)} results){c.RESET}"
                )

            user_input = input("Enter a number to run a use case, or enter a command (help for commands): ").strip()

            if not user_input:
                continue

            lowered = user_input.lower()
            if lowered in ("help", "h", "?"):
                self._show_help()
                skip_use_case_details = True
                continue

            if lowered == "summary":
                summary = True
                continue

            if lowered in ("full", "details"):
                summary = False
                continue

            if user_input.startswith("search "):
                self._search_use_cases(user_input[7:].strip(), False)
                continue

            if user_input.startswith("s "):
                self._search_use_cases(user_input[2:].strip(), False)
                continue

            if user_input.startswith("/"):
                self._search_use_cases(user_input[1:].strip(), True)
                continue

            if lowered in ("clear", "c"):
                self._search_use_cases(None, False)
                continue

            if lowered in ("exit", "quit"):
                print("Goodbye!")
                break

            try:
                selection = int(user_input)
            except ValueError:
                print(f"{c.RED}Invalid input. Please enter a number or a command (type 'help' for available commands).{c.RESET}")
                continue

            if 1 <= selection <= len(self.filtered_use_cases):
                self._invoke_use_case(selection)
            else:
                print(f"{c.RED}Invalid selection. Please enter a number between 1 and {len(self.filtered_use_cases)}.{c.RESET}")

    @staticmethod
    def _show_start_header() -> None:
        print()
        print(f"{c.BOLD}Aerospike Use Case Cookbook (Python SDK){c.RESET}")
        print("==============================")
        print()

    def _format_colors(self, text: str, color: str) -> str:
        return f"{color}{self._highlight_search_terms(text, color)}{c.RESET}"

    def _format_line(self, index_str: str, name_str: str, desc_str: str, fmt: str) -> str:
        return f"| {self._format_colors(index_str, fmt)} | {self._format_colors(name_str, fmt)} | {self._format_colors(desc_str, fmt)} |"

    @staticmethod
    def _form_tags_string(tags: list[str], color: str) -> str:
        parts = [f"{c.REVERSE}{tag}{c.RESET}{color}" for tag in tags]
        return "Tags: " + " ".join(parts)

    def _show_use_case_details(self, width: int, summary: bool) -> None:
        if not self.filtered_use_cases:
            print(f"{c.RED}No use cases found matching your search criteria.{c.RESET}")
            return

        column_spaces = 6
        column_chars = 4
        longest_name = max((len(uc.get_name()) for uc in self.filtered_use_cases), default=0)
        index_width = len(str(len(self.filtered_use_cases)))
        name_width = min(longest_name, 50)
        desc_width = max(width - index_width - name_width - column_chars - column_spaces, 10)

        index_header = self._pad_center("#", index_width)
        name_header = self._pad_center("Name", name_width)
        desc_header = self._pad_center("Summary" if summary else "Description", desc_width)

        horizontal_line = "-" * width
        print(horizontal_line)
        print(self._format_line(index_header, name_header, desc_header, c.BOLD))
        print(horizontal_line)

        for i, uc in enumerate(self.filtered_use_cases):
            index = str(i + 1).rjust(index_width)
            color = c.COLOR2 if i % 2 == 0 else c.COLOR1
            name = uc.get_name().ljust(name_width)
            wrapped_desc = self._wrap_text(self._form_use_case_text(uc, summary), desc_width)

            for j, desc_line in enumerate(wrapped_desc):
                padded = desc_line.ljust(desc_width)
                if j == 0:
                    print(self._format_line(index, name, padded, color))
                else:
                    print(self._format_line(" " * index_width, " " * name_width, padded, color))

            if uc.get_reference():
                print(self._format_line(" " * index_width, " " * name_width, " " * desc_width, color))
                print(self._format_line(
                    " " * index_width, " " * name_width,
                    f"See: {uc.get_reference()}".ljust(desc_width), color,
                ))
            if uc.get_tags():
                print(self._format_line(
                    " " * index_width, " " * name_width,
                    self._form_tags_string(uc.get_tags(), color).ljust(desc_width), color,
                ))
            print(horizontal_line)

    @staticmethod
    def _form_use_case_text(uc: UseCase, summary: bool) -> str:
        if summary:
            return uc.get_description().split(".")[0] + "."
        return uc.get_description()

    @staticmethod
    def _wrap_text(text: str, width: int) -> list[str]:
        lines: list[str] = []
        for paragraph in text.split("\n"):
            current_line = ""
            for word in paragraph.split():
                if len(current_line) + len(word) + 1 <= width:
                    current_line = f"{current_line} {word}".strip()
                else:
                    if current_line:
                        lines.append(current_line)
                        current_line = word
                    else:
                        lines.append(word[:width])
                        current_line = word[width:]
            if current_line:
                lines.append(current_line)
        return lines

    @staticmethod
    def _pad_center(text: str, width: int) -> str:
        left = max(0, (width - len(text)) // 2)
        right = max(0, width - left - len(text))
        return " " * left + text + " " * right

    def _highlight_search_terms(self, text: str, preserve_color: str) -> str:
        if not self.current_search_term or not self.current_search_term.strip():
            return text
        try:
            pattern = re.compile(
                self.current_search_term if self.is_regex_search else re.escape(self.current_search_term),
                re.IGNORECASE,
            )
        except re.error:
            pattern = re.compile(re.escape(self.current_search_term), re.IGNORECASE)
        return pattern.sub(lambda m: f"{c.HIGHLIGHT}{m.group()}{c.RESET}{preserve_color}", text)

    def _search_use_cases(self, search_term: str | None, use_regex: bool) -> None:
        self.current_search_term = search_term
        self.is_regex_search = use_regex

        if not search_term or not search_term.strip():
            self.filtered_use_cases = list(registry.get_all_use_cases())
            return

        try:
            pattern = re.compile(search_term if use_regex else re.escape(search_term), re.IGNORECASE)
        except re.error:
            pattern = re.compile(re.escape(search_term), re.IGNORECASE)

        self.filtered_use_cases = [uc for uc in registry.get_all_use_cases() if self._matches_search(uc, pattern)]

    @staticmethod
    def _matches_search(use_case: UseCase, pattern: "re.Pattern") -> bool:
        return bool(
            pattern.search(use_case.get_name())
            or pattern.search(use_case.get_description())
            or pattern.search(use_case.get_reference())
            or pattern.search(" ".join(use_case.get_tags()))
        )

    def _invoke_use_case(self, selection: int) -> None:
        uc = self.filtered_use_cases[selection - 1]
        self.executor.execute_use_case(uc, True, False, False)

    @staticmethod
    def _show_help() -> None:
        print("Help commands:")
        print("  help           - Show this help text")
        print("Display commands:")
        print("  summary        - Show only summary (one line) use case descriptions")
        print("  full / details - Show detailed use case descriptions")
        print("Search commands:")
        print("  search <term>  - Search for term in names, descriptions, and URLs")
        print("  s <term>       - Short form of search")
        print("  /<regex>       - Search using regular expression")
        print("  clear          - Clear search and show all use cases")
        print("Running a use case:")
        print("  <shown number> - Execute the use case with the specified number")
        print("                  (If the use case has parameters, you'll be prompted to configure them)")
        print("Exiting the program:")
        print("  exit / quit    - Exit the program")
        print()
